File: leetcode/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = root.val
        def traverse(node, presum):
            if node is None:
                return -sys.maxsize
            
            presum += node.val
            left = traverse(node.left, presum)
            right = traverse(node.right, presum)

            max_sum = node.val
            if left > presum:
                max_sum += left-presum
            if right > presum:
                max_sum += right-presum
            nonlocal res
            res = max(res, max_sum)

            return max(presum, left, right)
        
        traverse(root, 0)
        return res

File: leetcode/124. Binary Tree Maximum Path Sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        auto tp = helper(root);
        return get<0>(tp);
    }
    
private:
    tuple<int, int> helper(TreeNode* node) {
        if (node == nullptr) {
            return make_tuple(numeric_limits<int>::min(), numeric_limits<int>::min());
        }
        
        auto ltp = helper(node->left);
        auto rtp = helper(node->right);
        
        int max_path = node->val;
        if (get<1>(ltp) > 0) {
            max_path += get<1>(ltp);
        }
        if (get<1>(rtp) > 0) {
            max_path += get<1>(rtp);
        }
        
        max_path = max({max_path, get<0>(ltp), get<0>(rtp)});
        
        int max_branch = max({0, get<1>(ltp), get<1>(rtp)}) + node->val;
        
        return make_tuple(max_path, max_branch);
    }
};
